### PR TITLE
Refactoring parametrization

### PR DESCRIPTION
**Pull request by stefdoerr**
Okay so please keep in mind the scope of this PR. 
It's scope is to remove the old `PRM RTF AMBERPRM AMBERRTF FFMolecule FFType` and `FFEvaluate` classes. Not to improve functionality per se or improve code organization although where that happens naturally I have done it.

- [x] Replaced old classes with parmed
- [x] Parameters are printing in alphabetical order
- [x] CHARMM types have their typeindex written correclty
- [x] htmd/parameterize/test.py tests all run without error (but different+correct results obviously)
- [x] Need to fix QM tests, Dihedral tests
- [x] Remove ffmolecule.py
- [x] Update test-param test files.

### CODE REVIEW

**Comment by j3mdamas** [comment]
Thanks @stefdoerr, great work. Let's try to make it pass for May's stable release

File: htmd/parameterization/dihedral.py [hunk 1]
```diff
@@ -335,23 +287,33 @@ def _optimize_random_search(self, vector):
 
         return best_vector
 
+    def _updateDihedrals(self, parameters, keys, vector):
+        nparams = len(keys) * self.MAX_DIHEDRAL_MULTIPLICITY
+        assert vector.size == 2 * nparams + self.numDihedrals
+
+        for i, k in enumerate(keys):
+            for j, t in enumerate(parameters.dihedral_types[k]):
+                t.phi_k = vector[i*6+j]
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: You hard coding `self.MAX_DIHEDRAL_MULTIPLICITY`

File: htmd/parameterization/dihedral.py [hunk 2]
```diff
@@ -335,23 +287,33 @@ def _optimize_random_search(self, vector):
 
         return best_vector
 
+    def _updateDihedrals(self, parameters, keys, vector):
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: For consistency, the name should be `_vectorToParams`. So it is obvious, it is doing opposite transformation to  `_paramsToVector`

File: htmd/parameterization/dihedral.py [hunk 3]
```diff
@@ -236,32 +195,25 @@ def _objective(self, x, _):
...
         """
+        vector = []
+        for k in keys:
+            assert len(params.dihedral_types[k]) == self.MAX_DIHEDRAL_MULTIPLICITY
+            for term in params.dihedral_types[k]:
+                vector.append(term.phi_k)
+        for k in keys:
+            for term in params.dihedral_types[k]:
+                vector.append(term.phase)
+        for i in range(self.numDihedrals):
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: Why not just
```python
vector += [0] * self.numDihedrals
```

File: htmd/parameterization/dihedral.py [hunk 4]
```diff
@@ -366,16 +328,15 @@ def _fit(self):
         logger.info('Finished parameter optimization')
 
         # Update the target dihedral with the optimized parameters
-        self.parameters = self._vectorToParams(vector, self.parameters)
-        self.molecule._prm.updateDihedral(self.parameters)
+        self._updateDihedrals(self.parameters, self.fitdihkeys, vector)
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: This should return an updated copy:
```python
self.parameters = self._vectorToParams(vector, self.parameters, self.fitdihkeys)
```

**Comment by stefdoerr** [comment]
So one thing that bothers me a bit is this:
```
    mol = Molecule(molFile)
    mol = canonicalizeAtomNames(mol)
    mol, equivalents, all_dihedrals = getEquivalentsAndDihedrals(mol)  # Internally it guesses bonds, angles, dihedrals
```
I think it might make sense to move name canonicalization and bond, angle, dihedral guessing to the Molecule class and maybe make them also constructor options to avoid calling 3 different methods.

File: htmd/parameterization/dihedral.py (hunk 4 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: I don't use `vectorToParams` anymore. I use directly the `_updateDihedrals` method.

File: htmd/parameterization/dihedral.py (hunk 3 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: I was making the other two loops already so I thought I might as well make the third explicit

File: htmd/parameterization/dihedral.py (hunk 2 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Okay I will rename it.

File: htmd/parameterization/dihedral.py (hunk 1 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Thanks!

**Comment by stefdoerr** [comment]
Ok now officially all tests is `parameterize` folder run (with exception of `test.py` which runs but I have not updated the test files as I will do it as the last thing due to sheer amount of changes)

**Comment by j3mdamas** [comment]
But most tests are on `test.py`, no?

I'll to look at this by the end of the week.

**Comment by stefdoerr** [comment]
@j3mdamas Read again. The tests run fine and give good results. They fail because I sort parameters alphabetically and because I removed unused parameters from `frcmods` etc.

Okay. I finished the last of the tests. This PR is ready. Remaining questions are:
1. Whatever you find in the code
2. If I should simplify the `canonicalization` and the `angle/dihedral guessing` inside `Molecule` to reduce command spam

**Comment by stefdoerr** [comment]
On point number 2. I would propose doing it as follows:
```python
mol = Molecule(molfile, canonicalizeNames=True, guessAnglesDihedrals=True)
```
by adding two methods:
```python
mol.guessAnglesDihedrals()
mol.canonicalizeNames()
```
and adding them also as constructor options to avoid 3 commands.

**Comment by j3mdamas** [comment]
@stefdoerr so it passes locally but you haven't updated the files on the repo, is that it? I misunderstood.

**Comment by raimis** [comment]
@stefdoerr 
- Angle and dihedral guessing can go to `Molecule`, as you think it is better.
- Atom renaming will disappear, when I fix #524. So just leave it.

**Comment by stefdoerr** [comment]
Oh I see. I thought the renamed atoms were desired. Good, then I'll leave the command there for the moment and we remove it once this is completed.

@j3mdamas it runs locally. It doesn't pass because I have not substituted the files. But I checked the diff and the results are fine. Whatever, pointless distinctions once I update the test files.

**Comment by stefdoerr** [comment]
This PR is finished. @j3mdamas @raimis review it or pay the price

**Comment by j3mdamas** [comment]
Next Wednesday for me. @raimis, promise you don't commit more to parameterization files ;)

File: htmd/parameterization/test.py [hunk 5]
```diff
@@ -7,12 +7,77 @@
 import sys
 import shutil
 import unittest
-import numpy as np
 from subprocess import call
 
 from htmd.home import home
 from htmd.util import tempname
 
+
```
Comment (raimis) [inline htmd/parameterization/test.py]: This file cannot be changed, except the atom names.

File: htmd/qm/test.py [hunk 6]
```diff
@@ -14,7 +14,8 @@
 from htmd.queues.localqueue import LocalCPUQueue
 from htmd.queues.slurmqueue import SlurmQueue
 from htmd.queues.acecloudqueue import AceCloudQueue
-from htmd.parameterization.ffmolecule import FFMolecule, FFTypeMethod
+from htmd.parameterization.fftype import fftype, FFTypeMethod
```
Comment (raimis) [inline htmd/qm/test.py]: These are not needed here any more.

File: htmd/qm/psi4.py [hunk 7]
```diff
@@ -52,9 +52,13 @@ class Psi4(QMBase):
     Create an object of H2 molecule
     >>> import os
     >>> from htmd.home import home
-    >>> from htmd.parameterization.ffmolecule import FFMolecule, FFTypeMethod
+    >>> from htmd.parameterization.fftype import fftype, FFTypeMethod
```
Comment (raimis) [inline htmd/qm/psi4.py]: These are not need here anymore.

File: htmd/qm/fake.py [hunk 8]
```diff
@@ -325,6 +349,29 @@ def retrieve(self):
 
 
 if __name__ == '__main__':
+    import os
```
Comment (raimis) [inline htmd/qm/fake.py]: Clean this.

File: htmd/parameterization/test.py (hunk 5 above)
Comment (stefdoerr) [reply htmd/parameterization/test.py]: Sorry? I don't understand this comment

File: htmd/parameterization/esp.py [hunk 9]
```diff
@@ -201,24 +210,27 @@ def __init__(self):
 
         self._reciprocal_distances = None
 
+        self.equivalent_atom_groups = None
+        self.equivalent_group_by_atom = None
+        self.netcharge = None
```
Comment (raimis) [inline htmd/parameterization/esp.py]: This is wrong! The charge has to come from `qm_result` object.

File: htmd/parameterization/esp.py [hunk 10]
```diff
@@ -201,24 +210,27 @@ def __init__(self):
 
         self._reciprocal_distances = None
 
+        self.equivalent_atom_groups = None
```
Comment (raimis) [inline htmd/parameterization/esp.py]: Call equivalent atom detector internally, so we don't need this ugly attributes

File: htmd/parameterization/esp.py [hunk 11]
```diff
@@ -52,9 +52,15 @@ class ESP:
...
     >>> from htmd.home import home
-    >>> from htmd.parameterization.ffmolecule import FFMolecule, FFTypeMethod
+    >>> from htmd.parameterization.fftype import fftype, FFTypeMethod
+    >>> from htmd.parameterization.util import getEquivalentsAndDihedrals, canonicalizeAtomNames
+    >>> from htmd.molecule.molecule import Molecule
     >>> molFile = os.path.join(home('test-qm'), 'H2O.mol2')
-    >>> mol = FFMolecule(molFile, method=FFTypeMethod.GAFF2)
+    >>> mol = Molecule(molFile)
+    >>> mol = canonicalizeAtomNames(mol)
+    >>> parameters, mol = fftype(mol, method=FFTypeMethod.GAFF2)
```
Comment (raimis) [inline htmd/parameterization/esp.py]: No atom renaming and atom typing needed here

File: htmd/parameterization/esp.py [hunk 12]
```diff
@@ -298,17 +310,24 @@ def run(self):
...
-        from htmd.parameterization.ffmolecule import FFMolecule, FFTypeMethod
+        from htmd.parameterization.fftype import FFTypeMethod, fftype
+        from htmd.parameterization.util import getEquivalentsAndDihedrals, canonicalizeAtomNames
+        from htmd.molecule.molecule import Molecule
 
         molFile = os.path.join(home('test-param'), 'H2O2.mol2')
-        self.mol = FFMolecule(molFile, method=FFTypeMethod.GAFF2)
+        mol = Molecule(molFile)
+        mol = canonicalizeAtomNames(mol)
+        mol, equivalents, all_dihedrals = getEquivalentsAndDihedrals(mol)
```
Comment (raimis) [inline htmd/parameterization/esp.py]: No atom renaming and atom typing needed here

File: htmd/parameterization/dihedral.py [hunk 13]
```diff
@@ -154,23 +117,19 @@ def _setup(self):
...
 
         # Get equivalent dihedral atom indices
         self._equivalent_indices = []
-        for dihedral, name in zip(self.dihedrals, self._names):
-            for equivlantet_diherals in self.molecule._parameterizable_dihedrals:
-                if equivlantet_diherals[0] == dihedral:
-                    self._equivalent_indices.append(equivlantet_diherals)
+        for idihed, dihedral in enumerate(self.dihedrals):
+            found = False
+            for rotatableDihedral in self._rotatable_dihedrals:
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: `rotatable`--> `parameterizable`. I'm trying to make this consistent thought the code.

File: htmd/parameterization/cli.py [hunk 14]
```diff
@@ -153,26 +177,39 @@ def main_parameterize(arguments=None):
...
-        mol = FFMolecule(args.filename, method=methods[0], netcharge=args.charge, rtf=rtf, prm=prm, qm=qm,
-                         outdir=args.outdir)
-        print('\n === Parameterizable dihedral angles of %s ===\n' % args.filename)
+        print('\n === Parameterizable dihedral angles of {} ===\n'.format(args.filename))
         with open('torsions.txt', 'w') as fh:
-            for dihedral in mol.getParameterizableDihedrals():
-                dihedral_name = '-'.join(mol.name[list(dihedral)])
-                print('  '+dihedral_name)
-                fh.write(dihedral_name+'\n')
+            for dih in all_dihedrals:
```
Comment (raimis) [inline htmd/parameterization/cli.py]: `dih` --> `dihedral`
`dihname` --> `dihedral_name`

File: htmd/parameterization/cli.py [hunk 15]
```diff
@@ -153,26 +177,39 @@ def main_parameterize(arguments=None):
...
 
+    # Set up the QM object
+    qm.theory = args.theory
+    qm.basis = args.basis
+    qm.solvent = args.environment
+    qm.queue = queue
+    qm.netcharge = netcharge
+
+    # Select which dihedrals to fit
+    fit_dihedrals = [list(dih[0]) for dih in all_dihedrals]
```
Comment (raimis) [inline htmd/parameterization/cli.py]: `fit` --> `parameterizable`

File: htmd/parameterization/detect.py [hunk 16]
```diff
@@ -145,6 +145,9 @@ def _getMethylGraph():
 
     return methyl
 
+def connected_component_subgraphs(graph):
+    return (graph.subgraph(c).copy() for c in nx.connected_components(graph))
```
Comment (raimis) [inline htmd/parameterization/detect.py]: `c` --> `component`

File: htmd/parameterization/detect.py [hunk 17]
```diff
@@ -262,66 +265,74 @@ def detectParameterizableDihedrals(molecule):
...
+    >>> mol = Molecule(molFile, guess=('bonds', 'angles', 'dihedrals'))
     >>> detectParameterizableDihedrals(mol)
-    [[(2, 1, 0, 4), (1, 2, 3, 9)], [(0, 1, 2, 3)]]
+    [[(0, 1, 2, 3)], [(1, 2, 3, 9), (2, 1, 0, 4)]]
 
     Find the parameterizable dihedrals of ethanolamine
     >>> molFile = os.path.join(home('test-param'), 'ethanolamine.mol2')
-    >>> mol = FFMolecule(molFile, method=FFTypeMethod.GAFF2) # Does not work with CGenFF
+    >>> mol = Molecule(molFile, guess=('bonds', 'angles', 'dihedrals'))
+    >>> mol = canonicalizeAtomNames(mol)
```
Comment (raimis) [inline htmd/parameterization/detect.py]: No renaming needed here.

File: htmd/parameterization/detect.py [hunk 18]
```diff
@@ -262,66 +265,74 @@ def detectParameterizableDihedrals(molecule):
...
+    >>> prm, mol = fftype(mol, method=FFTypeMethod.GAFF2) # Does not work with CGenFF
```
Comment (raimis) [inline htmd/parameterization/detect.py]: No type guessing needed here

File: htmd/parameterization/detect.py [hunk 19]
```diff
@@ -262,66 +265,74 @@ def detectParameterizableDihedrals(molecule):
...
-    [[(1, 0, 6, 12), (1, 0, 6, 13), (5, 0, 6, 12), (5, 0, 6, 13)], [(0, 6, 12, 16), (0, 6, 12, 17), (0, 6, 13, 14), (0, 6, 13, 15)]]
+    [[(0, 6, 12, 16), (0, 6, 12, 17), (0, 6, 13, 14), (0, 6, 13, 15)], [(1, 0, 6, 12), (1, 0, 6, 13), (5, 0, 6, 12), (5, 0, 6, 13)]]
 
     # Check if the atom swapping does not affect results
 
     Find the parameterizable dihedrals of chlorethene
     >>> molFile = os.path.join(home('test-param'), 'chlorethene_1.mol2')
-    >>> mol = FFMolecule(molFile)
+    >>> mol = Molecule(molFile, guess=('bonds', 'angles', 'dihedrals'))
+    >>> mol = canonicalizeAtomNames(mol)
```
Comment (raimis) [inline htmd/parameterization/detect.py]: No renaming and type guessing needed here

File: htmd/parameterization/detect.py [hunk 20]
```diff
@@ -262,66 +265,74 @@ def detectParameterizableDihedrals(molecule):
...
+    >>> prm, mol = fftype(mol, method=FFTypeMethod.CGenFF_2b6)
     >>> detectParameterizableDihedrals(mol)
     [[(2, 1, 0, 4), (2, 1, 0, 5)]]
 
     Find the parameterizable dihedrals of chlorethene (with swapped atoms)
     >>> molFile = os.path.join(home('test-param'), 'chlorethene_2.mol2')
-    >>> mol = FFMolecule(molFile)
+    >>> mol = Molecule(molFile, guess=('bonds', 'angles', 'dihedrals'))
+    >>> mol = canonicalizeAtomNames(mol)
```
Comment (raimis) [inline htmd/parameterization/detect.py]: No renaming and type guessing needed here

File: htmd/parameterization/detect.py [hunk 21]
```diff
@@ -351,9 +362,9 @@ def detectParameterizableDihedrals(molecule):
     equivalent_dihedrals = OrderedDict()
     for dihedral, groups in zip(dihedrals, dihedral_groups):
         dihedral, groups = (dihedral[::-1], groups[::-1]) if groups[::-1] < groups else (dihedral, groups)
-        equivalent_dihedrals[groups] = equivalent_dihedrals.get(groups, []) + [dihedral]
-    equivalent_dihedrals = list(equivalent_dihedrals.values())
-
+        equivalent_dihedrals[groups] = sorted(equivalent_dihedrals.get(groups, []) + [dihedral])
+    equivalent_dihedrals = sorted(equivalent_dihedrals.values())
+    # print(equivalent_dihedrals)
```
Comment (raimis) [inline htmd/parameterization/detect.py]: Clean

File: htmd/parameterization/cli.py [hunk 22]
```diff
@@ -207,7 +245,8 @@ def main_parameterize(arguments=None):
         # Minimize molecule
         if args.minimize:
             print('\n == Minimizing ==\n')
-            mol.minimize()
+            # molFF.minimize()
```
Comment (raimis) [inline htmd/parameterization/cli.py]: Clean

File: htmd/parameterization/cli.py [hunk 23]
```diff
@@ -218,26 +257,18 @@ def main_parameterize(arguments=None):
...
-
-                for aton_index in range(mol.numAtoms):
-                    if mol.name[aton_index] == fixed_atom_name:
-                        fixed_atom_indices.append(aton_index)
-                        logger.info('Charge of atom %s is fixed to %f' % (fixed_atom_name, mol.charge[aton_index]))
+            fixed_atom_indices = getFixedChargeAtomIndices(mol, args.fix_charge)
 
             # Fit ESP charges
-            _, qm_dipole = mol.fitCharges(fixed=fixed_atom_indices)
+            # _, qm_dipole = molFF.fitCharges(fixed=fixed_atom_indices)
```
Comment (raimis) [inline htmd/parameterization/cli.py]: Clean

File: htmd/parameterization/cli.py [hunk 24]
```diff
@@ -218,26 +257,18 @@ def main_parameterize(arguments=None):
...
+            mol, _, esp_charges, qm_dipole = fitCharges(mol, qm, equivalents, netcharge, args.outdir, fixed=fixed_atom_indices)
 
             # Copy the new charges to the original molecule
-            mol_orig.charge[:] = mol.charge
+            # mol_orig.charge[:] = molFF.charge
```
Comment (raimis) [inline htmd/parameterization/cli.py]: Clean

File: htmd/parameterization/cli.py [hunk 25]
```diff
@@ -251,28 +282,25 @@ def main_parameterize(arguments=None):
...
+            # Invent new atom types for dihedral atoms
+            mol, originaltypes = inventAtomTypes(mol, fit_dihedrals, equivalents)
+            parameters = recreateParameters(mol, originaltypes, parameters)
+            parameters = createMultitermDihedralTypes(parameters)
+            if isinstance(qm, FakeQM2):
+                qm._parameters = parameters
 
             # Fit the parameters
-            mol.fitDihedrals(dihedrals, args.optimize_dihedral)
+            # molFF.fitDihedrals(fit_dihedrals, args.optimize_dihedral)
```
Comment (raimis) [inline htmd/parameterization/cli.py]: Clean

File: htmd/parameterization/dihedral.py [hunk 26]
```diff
@@ -54,6 +54,7 @@ def __init__(self):
         self.zeroed_parameters = False
 
         self.parameters = None
+        self.fitdihkeys = None
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: `fitdihkeys` -- > `_parameterizable_dihedral_atomtypes`

File: htmd/parameterization/dihedral.py [hunk 27]
```diff
@@ -71,53 +72,15 @@ def __init__(self):
         self._all_target_energies = None
         self._angle_values_rad = None
 
+        self.parmedMode = False
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: What is this for?

File: htmd/parameterization/dihedral.py [hunk 28]
```diff
@@ -71,53 +72,15 @@ def __init__(self):
...
+        self._prm = None
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: And this?

File: htmd/parameterization/dihedral.py [hunk 29]
```diff
@@ -71,53 +72,15 @@ def __init__(self):
...
+        self._rotatable_dihedrals = None
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: Call the parameterizable dihedral detector internally, so no need to pass this from outside.

File: htmd/parameterization/dihedral.py [hunk 30]
```diff
@@ -236,32 +197,25 @@ def _objective(self, x, _):
 
         return rmsd
 
-    def _paramsToVector(self, params):
+    def _paramsToVector(self, params, keys):
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: What the `key` stands for? Dihedral name, dihedral atom type names?

Comment (raimis) [reply htmd/parameterization/dihedral.py]: OK! Actually, it is `dihedral_atomtypes`.

File: htmd/parameterization/dihedral.py [hunk 31]
```diff
@@ -335,23 +289,33 @@ def _optimize_random_search(self, vector):
 
         return best_vector
 
+    def _vectorToParams(self, parameters, keys, vector):
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: `keys` --> `dihedral_atomtypes`

File: htmd/parameterization/dihedral.py [hunk 32]
```diff
@@ -532,35 +455,11 @@ def test_getValidQMResults(self):
         self.assertEqual(len(self.df._getValidQMResults()[0]), 17)
 
     def test_getBounds(self):
-
         self.df.dihedrals = [[0, 1, 2, 3]]
         lower_bounds, upper_bounds = self.df._getBounds()
         self.assertListEqual(list(lower_bounds), [0.,  0.,  0.,  0.,  0.,  0.,   0.,   0.,   0.,   0.,   0.,   0., -10.])
         self.assertListEqual(list(upper_bounds), [10., 10., 10., 10., 10., 10., 360., 360., 360., 360., 360., 360.,  10.])
 
-    def test_paramsToVector(self):
```
Comment (raimis) [inline htmd/parameterization/dihedral.py]: The methods are not removed, so the tests have to be preserved.

File: htmd/parameterization/dihedral.py (hunk 26 above)
Comment (raimis) [reply htmd/parameterization/dihedral.py]: And move it down

File: htmd/parameterization/test.py (hunk 5 above)
Comment (raimis) [reply htmd/parameterization/test.py]: Do not change `htmd/parameterization/test.py` in this PR, except for the atom names! It seems your are pick up the old code.

File: htmd/parameterization/cli.py (hunk 14 above)
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Changed

File: htmd/parameterization/cli.py (hunk 15 above)
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Done

File: htmd/parameterization/cli.py (hunk 22 above)
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Cleaned

File: htmd/parameterization/cli.py (hunk 23 above)
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Cleaned

File: htmd/parameterization/cli.py (hunk 24 above)
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Cleaned

File: htmd/parameterization/cli.py (hunk 25 above)
Comment (stefdoerr) [reply htmd/parameterization/cli.py]: Cleaned

File: htmd/parameterization/detect.py (hunk 16 above)
Comment (stefdoerr) [reply htmd/parameterization/detect.py]: done

File: htmd/parameterization/detect.py (hunk 17 above)
Comment (stefdoerr) [reply htmd/parameterization/detect.py]: Removed

Comment (stefdoerr) [reply htmd/parameterization/detect.py]: Actually no. Fails if removed

File: htmd/parameterization/detect.py (hunk 18 above)
Comment (stefdoerr) [reply htmd/parameterization/detect.py]: Ah ok removed both and it works.

File: htmd/parameterization/detect.py (hunk 19 above)
Comment (stefdoerr) [reply htmd/parameterization/detect.py]: rm-ed

File: htmd/parameterization/detect.py (hunk 20 above)
Comment (stefdoerr) [reply htmd/parameterization/detect.py]: rm-ed

File: htmd/parameterization/detect.py (hunk 21 above)
Comment (stefdoerr) [reply htmd/parameterization/detect.py]: cleaned

File: htmd/parameterization/dihedral.py (hunk 26 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Renamed and moved

File: htmd/parameterization/dihedral.py (hunk 27 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: debugging. removed

File: htmd/parameterization/dihedral.py (hunk 28 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Old stuff. removed

File: htmd/parameterization/dihedral.py (hunk 29 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Withe the current design of parameterize I don't agree to this change, as it leads to calculation duplication. We will need to discuss the class design and if we end up making the classes stand-alone I agree to remove it.

Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: But for a following PR. Not this one

File: htmd/parameterization/dihedral.py (hunk 13 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: renamed

File: htmd/parameterization/dihedral.py (hunk 30 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Renamed

File: htmd/parameterization/dihedral.py (hunk 31 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Changed

File: htmd/parameterization/dihedral.py (hunk 32 above)
Comment (stefdoerr) [reply htmd/parameterization/dihedral.py]: Readded and updated for the new code

File: htmd/parameterization/esp.py (hunk 11 above)
Comment (stefdoerr) [reply htmd/parameterization/esp.py]: Thanks. Removed

File: htmd/parameterization/esp.py (hunk 10 above)
Comment (stefdoerr) [reply htmd/parameterization/esp.py]: See other comment on class redesign. I'll leave it for the moment like this. We can discuss it tomorrow

File: htmd/parameterization/esp.py (hunk 12 above)
Comment (stefdoerr) [reply htmd/parameterization/esp.py]: Removed

File: htmd/parameterization/fftype.py [hunk 33]
```diff
@@ -21,57 +26,39 @@ class FFTypeMethod(Enum):
     GAFF2 = 1002
 
 
-class FFType:
-    """
```
[the remark this answers is not in the document]
Comment (stefdoerr) [reply htmd/parameterization/fftype.py]: Readded and updated

**Comment by stefdoerr** [comment]
@raimis I made the changes. But one thing. What am I supposed to do with test_multiplicity in `qm/test.py`? Since you told me to remove the `netcharge` argument from `QM` I cannot test different charges anymore. Should I assign the charge to a random `Molecule` atom?

File: htmd/parameterization/esp.py (hunk 9 above)
Comment (stefdoerr) [reply htmd/parameterization/esp.py]: You will need to help me with this. I only see some `mulliken` charges in `QMResult`. What should I do with them?

**Comment by j3mdamas** [comment]
I'm going to sign out on reviewing this, I think @raimis is already giving you a sufficient hard time :P

Some small comments:
* It's a lot of refactoring work and juggling around code. Kudos.
* I think that replacing faulty parameter structures (and etc) with parmed "objects" should be beneficial in the future (as long as that project stands)
* I'm still clueless on the advantages of killing FFMolecule and its methods for a bunch of functions on utils that take Molecules and spit out Molecules (many times), but I guess that's my ignorance. Maybe you need to explain me that better.
* I think one of the advantages of a Molecule with FF parameters would be to be able to do "physics-based" operations on a high-level (for example, rotate a torsion and evaluate energy) that could be useful further ahead, as Gianni once mentioned, but maybe that's dropped (?)

Anyway, I'll leave you to it.

**Comment by stefdoerr** [comment]
Also the thing about the `mulliken` charges which I don't know how to handle

**Comment by stefdoerr** [comment]
Also in updating the `psi4.in` I seem to have messed up some `PSI4` settings, probably memory or cpu, and glycol tests take ages to finish. You will need to help me with this as well.

**Comment by j3mdamas** [comment]
Just a question: doesn't this depend on #564?

**Comment by stefdoerr** [comment]
Nah I forgot to kill that PR/branch. I have merged it's changes in the new
one

**Comment by j3mdamas** [comment]
I'm glad I'm on `rel-1.12.x` branch only :D

**Comment by raimis** [comment]
Thanks @stefdoerr for refactoring a lot of messy code!

**Comment by j3mdamas** [comment]
yes, kudos! But now, after tearing apart, we need to give it some structure ;)

**Pull request merged**